### PR TITLE
[graph] Skip backwarding bug fix + memory optimize

### LINKED ISSUE #1664: ENABLE_TEST=false does not work
ENABLE_TEST=false does not work on some models. 

Error is requesting gradient for non trainable tensors.

I think if there is a intermediat non trainable tensor, it seems affecting it.

### LINKED ISSUE #1686: [graph] Skipping backwarding is faulty
The current code for skipping backwarding assumes a linear model graph, and skips the backwarding only for the layers at the start of the model which dont need backwardng. However, this does not support for complicated graphs where intermediate layers also dont need backwarding.

### PR DESCRIPTION
**Pull request by kparichay**
Current implementation assumes that the graph is always linear and skips
the backwarding for contiguous set of layers at the start of the model
which should not do the backwarding.

This patch makes a fix to this issue but allowing layers in the middle
of the sorted graph to skip backwarding.

Resolves #1686
Resolves #1664

Note: this will provide minor speedup for complex models where some more layers backwarding can be skipped.

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Comment by kparichay** [comment]
V2:
- apply bug fix for lstmcell
- Currently, the memory was allocated assuming all the nodes were running
backwarding, and all the layers are trainable. But that is not true.
This patch applies the corresponding optimization and allocates memory
corresponding to the functions which will run on these nodes.

File: nntrainer/graph/network_graph.cpp [hunk 1]
```diff
@@ -313,6 +313,12 @@ int NetworkGraph::checkInitializedGraph() {
     if (lnode->getTrainable() ||
         must_support_backwarding.find(lnode->getName()) !=
           must_support_backwarding.end()) {
+      lnode->needsCalcGradient(true);
+#ifdef ENABLE_TEST
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:317]: This test flag is likely to be broken. Could we stick to less optimized version, and later optimize + fix test along instead?

File: nntrainer/graph/network_graph.cpp [hunk 2]
```diff
@@ -313,6 +313,12 @@ int NetworkGraph::checkInitializedGraph() {
...
+      if (lnode->supportBackwarding() && !optimize_memory) {
+        lnode->needsCalcDerivative(true);
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:319]: This function name suggest to check and not to touch the node property. If it is now, how about split this to different function or change the function name?

File: nntrainer/graph/network_graph.cpp [hunk 3]
```diff
@@ -313,6 +313,12 @@ int NetworkGraph::checkInitializedGraph() {
     if (lnode->getTrainable() ||
         must_support_backwarding.find(lnode->getName()) !=
           must_support_backwarding.end()) {
+      lnode->needsCalcGradient(true);
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:316]: If we are going to do this, I think layerNode::calcGradient should not implicitly make calcGradient noop.

below is related snippet.
```cpp
  if (getTrainable())
    layer->calcGradient(*run_context);
```

File: nntrainer/graph/network_graph.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:319]: There are two needsCalcDerivative(). One is a getter, one is a setter. Do you want different names for the two?

File: nntrainer/graph/network_graph.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:316]: updated  `if (getTrainable())` to `if (needs_calc_gradient)`.

File: nntrainer/graph/network_graph.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:319]: no it's about `checkInitializedGraph()` seems not clear to me.

it sounds like a sanity check but actually writing something

File: nntrainer/graph/network_graph.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:317]: In general, we want to run optimized version where we want to skip running the nodes for backwarding where its not needed, and so not allocate memory for it either.
With ENABLE_TEST, we disable such optimizations. The issue is not the memory optimization here but performance as calcDerivative runs for nodes which ideally should not.

File: nntrainer/graph/network_graph.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:319]: I understand, I will split it into two functions.

File: nntrainer/graph/network_graph.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:317]: I agree, my concern is that by we have `#ifndef ENABLE_TEST` part is never being tested, so from #1664 there is symptoms of a bug (android build is enabling this flag to true for now).

So I am suggesting to comment this out, remove #ifdef flag and make some changes to the test while switching to the optimized version.

Comment (jijoongmoon) [reply nntrainer/graph/network_graph.cpp:317]: > performance as calcDerivative runs for nodes which ideally should not.

I wonder which nodes are? You mean the feature extract cases which only support forwarding? If it is, then it's ok.
But other than that... None of the cases comes to my mind.

File: nntrainer/layers/layer_node.h [hunk 4]
```diff
@@ -661,11 +661,47 @@ class LayerNode final : public ml::train::Layer, public GraphNode {
    */
   std::unique_ptr<LayerNode> cloneConfiguration();
 
+  /**
+   * @brief Set if the layer needs to do derivative calculation
+   *
+   * @param nb true if the layer needs to do backwarding, else false
+   */
+  void needsCalcDerivative(bool nb) {
```
Comment (jijoongmoon) [inline nntrainer/layers/layer_node.h:669]: so.. we have..
1. supportBackwarding
2. needsBackwarding --> needsCalcDerivative & needsCalCGradient?

File: nntrainer/graph/network_graph.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:317]: If the first few layers are non-trainable but support backward (for example, input layer followed by multi-out layer), the multi-out layer's calcDerivative should not be called. 
This patch ensures that multi-out layer's calcDerivative is not called, and the corresponding memory is also not allocated.

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:317]: There are only three locations where ENABLE_TEST is being used (2 of which are from this PR itself). I will check the issue from 1664 and see what is the problem, but ideally, we need to run some tests with ENABLE_TEST as well.

File: nntrainer/layers/layer_node.h (hunk 4 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.h:669]: supportBackwarding is acutally supportCalcDerivative (didn't change this as its easier to understand for the layer developers), but internally, needsBackwarding is created and split into two.

Comment (jijoongmoon) [reply nntrainer/layers/layer_node.h:669]: Thanks for the explanation.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/graph/network_graph.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:317]: Marking resolves #1664 as this patch solves that issue.

File: nntrainer/graph/network_graph.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:319]: renamed to `markNodesForBackwarding`.

**Pull request merged**